### ISSUE TITLE
Calling `MonoidalCategoriesTest` with `only_primitive_operations := true` fails

### ISSUE THREAD

**Issue by TKuh** [homalg-project/CAP_project#886]
I'm trying to add the additive monoidal tests to CategoryOfRows as suggested by @zickgraf in https://github.com/homalg-project/CAP_project/pull/882#issuecomment-1108195977 and also wanted to add calls with the option `only_primitive_operations` as done in other variants of the tests (e.g. [here](https://github.com/homalg-project/FinSetsForCAP/blob/master/tst/CartesianClosedOpposite.tst#L25-L26)). However, when I put such calls with the option `only_primitive_operations := true` first, the following happens:

```gap
gap> LoadPackage( "FreydCategoriesForCAP" );
gap> Q := HomalgFieldOfRationals();;
gap> rows := CategoryOfRows( Q );;
gap> a := 2 / rows;;
gap> b := 3 / rows;;
gap> c := 6 / rows;;
gap> d := 4 / rows;;

gap> alpha := CategoryOfRowsMorphism( a, HomalgMatrix( [ 2 .. 7 ], RankOfObject( a ), RankOfObject( b ), Q ), b );;
gap> beta := CategoryOfRowsMorphism( c, HomalgMatrix( [ 8 .. 31 ], RankOfObject( c ), RankOfObject( d ), Q ), d );;

gap> MonoidalCategoriesTest( rows, a, b, c, alpha, beta : only_primitive_operations := true );;
gap> MonoidalCategoriesTest( rows, a, b, c, alpha, beta );;

gap> a := 0 / rows;;
gap> b := 3 / rows;;

gap> alpha := ZeroMorphism( a, b );;
gap> 
gap> MonoidalCategoriesTest( rows, a, b, c, alpha, beta : only_primitive_operations := true );;
Error, the number of rows has to be equal to the rank of the source at /home/tom/.gap/pkg/CAP_project/FreydCategoriesForCAP/gap/CategoryOfRows.gi:192 called from
CategoryOfRowsMorphism( cat, s, KroneckerMat( UnderlyingMatrix( alpha ), UnderlyingMatrix( beta ) ), r 
 ) at /home/tom/.gap/pkg/CAP_project/FreydCategoriesForCAP/gap/CategoryOfRows.gi:969 called from
CallFuncList( func_to_install, arg ) at /home/tom/.gap/pkg/CAP_project/CAP/gap/InstallAdds.gi:502 called from
CallFuncList( func, arg ) at /home/tom/.gap/pkg/homalg_project/ToolsForHomalg/gap/CachingObjects.gi:771 called from
TensorProductOnMorphismsWithGivenTensorProducts( OppositeCategory( cat ), ObjectDatum( cat, s ), 
 MorphismDatum( cat, alpha ), MorphismDatum( cat, beta ), ObjectDatum( cat, r ) ) at stream:7 called from
CallFuncList( func_to_install, arg ) at /home/tom/.gap/pkg/CAP_project/CAP/gap/InstallAdds.gi:502 called from
...  at *stdin*:20

brk> Display( source );
A row module over Q of rank 12
brk> Display( range );
A row module over Q of rank 0
brk> Display( homalg_matrix );
(an empty 0 x 12 matrix)

```
Apparently, `source` and `range` are mistaken, but I struggle to get any more meaningful information from the above break loop. :/ Could it be the caching? If I make the calls with `only_primitive_operations := true` only after a call without this option, I receive no errors:

```gap
gap> LoadPackage( "FreydCategoriesForCAP" );
gap> Q := HomalgFieldOfRationals();;
gap> rows := CategoryOfRows( Q );;
gap> a := 2 / rows;;
gap> b := 3 / rows;;
gap> c := 6 / rows;;
gap> d := 4 / rows;;
gap> 
gap> alpha := CategoryOfRowsMorphism( a, HomalgMatrix( [ 2 .. 7 ], RankOfObject( a ), RankOfObject( b ), Q ), b );;
gap> beta := CategoryOfRowsMorphism( c, HomalgMatrix( [ 8 .. 31 ], RankOfObject( c ), RankOfObject( d ), Q ), d );;
gap> 
gap> MonoidalCategoriesTest( rows, a, b, c, alpha, beta );;
gap> MonoidalCategoriesTest( rows, a, b, c, alpha, beta : only_primitive_operations := true );;
gap> 
gap> a := 0 / rows;;
gap> b := 3 / rows;;
gap> 
gap> alpha := ZeroMorphism( a, b );;
gap>
gap> MonoidalCategoriesTest( rows, a, b, c, alpha, beta );;
gap> MonoidalCategoriesTest( rows, a, b, c, alpha, beta : only_primitive_operations := true );;
gap> 

```

**Comment by zickgraf** [comment]
I'm not sure, but for me this looks like `TensorProductOnMorphismsWithGivenTensorProducts` has to swap source and range for the opposite category but does not. Could this be the problem?

**Comment by zickgraf** [comment]
Btw, do you know the `Where` and `DownEnv` and `UpEnv` commands? If not, you can read chapters 6.4 and 6.5 in the GAP documentation. The call stack in your example is quite shallow (only three CAP operations, you can ignore all the `CallFuncList`s), so you can check at which point things go wrong.

**Comment by TKuh** [comment]
> I'm not sure, but for me this looks like `TensorProductOnMorphismsWithGivenTensorProducts` has to swap source and range for the opposite category but does not. Could this be the problem?

Thank you very much for your fast reply, that has been the culprit! :) However, adding `only_primitve_operations := true` everywhere increases the runtime of the opposite tests to 666239 ms, so I'll drop my plans. :D But everything works at least locally.

> Btw, do you know the `Where` and `DownEnv` and `UpEnv` commands? If not, you can read chapters 6.4 and 6.5 in the GAP documentation. The call stack in your example is quite shallow (only three CAP operations, you can ignore all the `CallFuncList`s), so you can check at which point things go wrong.

Yes, I also read about it and went up in the environment to here
`TensorProductOnMorphismsWithGivenTensorProducts( OppositeCategory( cat ), ObjectDatum( cat, s ), 
 MorphismDatum( cat, alpha ), MorphismDatum( cat, beta ), ObjectDatum( cat, r ) ) at stream:7`

but then I struggled to get the above arguments and couldn't find out where this call to `TensorProductOnMorphism` occurs. And I didn't think the problem could be in the method records (again ...). Still, thanks for taking your time to point me there!

**Comment by mohamed-barakat** [comment]
In this context, we should also think about how to use `TerminalCategoryWithMultipleObjects` in combination with `LazyCategory` to *automatically* test the type-correctness of Add-methods/derived methods, maybe even independent of Fabian's type system.

**Comment by zickgraf** [comment]
> In this context, we should also think about how to use `TerminalCategoryWithMultipleObjects` in combination with `LazyCategory` to _automatically_ test the type-correctness of Add-methods/derived methods, maybe even independent of Fabian's type system.

I don't think this would have helped here: we were trying to create a non-well-defined morphism, `LazyCategory` would not prevent this.

**Comment by mohamed-barakat** [comment]
> I don't think this would have helped here: we were trying to create a non-well-defined morphism, `LazyCategory` would not prevent this.

The suggestion is not to address the problem here. I just was too lazy to open a new issue :)

**Comment by TKuh** [comment]
> > However, adding `only_primitve_operations := true` everywhere increases the runtime of the opposite tests to 666239 ms, so I'll drop my plans. :D
> 
> If you have the time and motivation, you could check if this is slow due to (un)wrapping or due to some different derivation path. The latter case would be something we should look at.

Calling `ClosedMonoidalCategoriesTest( ... ) ` and `ClosedMonoidalCategoriesTest( ... : only_primitive_operations )` yields 1.909 minutes vs 1.884 minutes. I guess the increase in runtime is because every test function is called twice (with and without `only_primitive_operations`) and again the only operations that consume time are the compatibility morphisms. So I suggest, I add the calls with `only_primitive_operations` but slightly reduce the dimensions of the vector spaces, so that the compatibility morphisms are fast. Locally for me, everything works fast again then.

**Comment by zickgraf** [comment]
> Calling `ClosedMonoidalCategoriesTest( ... ) ` and `ClosedMonoidalCategoriesTest( ... : only_primitive_operations )` yields 1.909 minutes vs 1.884 minutes.

So basically they have the same runtime, correct?

> I guess the increase in runtime is because every test function is called twice (with and without `only_primitive_operations`) and again the only operations that consume time are the compatibility morphisms.

Does this refer to the "666239 ms" you mentioned earlier? What's the baseline (i.e. the runtime before your changes/additions) for this number? And how does this number relate to the 1.909 minutes?

Sorry if I don't get obvious things, I'm a bit confused by what all these numbers actually refer to :D

**Comment by TKuh** [comment]
I reran everything to make the comparisons clearer and found out the following:

Currently:
- testing: /home/tom/.gap/pkg/CAP_project/FreydCategoriesForCAP/tst/TensorCorrespondenceRowsAndOppositeOfRows.tst
    6308 ms (2801 ms GC) and 1.11GB allocated for TensorCorrespondenceRowsAndOppositeOfRows.tst

Adding calls with `only_primitive_operations` before the calls without this option (https://github.com/TKuh/CAP_project/commit/768d00aa48359956022c01602a040317a6176d29):
- testing: /home/tom/.gap/pkg/CAP_project/FreydCategoriesForCAP/tst/TensorCorrespondenceRowsAndOppositeOfRows.tst
  515724 ms (6602 ms GC) and 2.86GB allocated for TensorCorrespondenceRowsAndOppositeOfRows.tst

Adding calls with `only_primitive_operations` after the calls without this option (https://github.com/TKuh/CAP_project/commit/46d946ad3a6c0a4af1b42e6bb7012ecf8d83513c):
- testing: /home/tom/.gap/pkg/CAP_project/FreydCategoriesForCAP/tst/TensorCorrespondenceRowsAndOppositeOfRows.tst
    6111 ms (2495 ms GC) and 1.14GB allocated for TensorCorrespondenceRowsAndOppositeOfRows.tst

Executing only a single test function and measuring its time:

```gap
gap> LoadPackage("freyd");
gap> Q := HomalgFieldOfRationals();;
gap> rows := CategoryOfRows( Q );;
gap> u := TensorUnit( rows );;
gap> 
gap> a := 2 / rows;;
gap> b := 3 / rows;;
gap> c := 6 / rows;;
gap> d := 4 / rows;;
gap> 
gap> a_tensor_b := TensorProduct( a, b );;
gap> c_tensor_d := TensorProduct( c, d );;
gap> 
gap> hom_ab := InternalHom( a, b );;
gap> hom_cd := InternalHom( c, d );;
gap> 
gap> alpha := CategoryOfRowsMorphism( a, HomalgMatrix( [ 2 .. 7 ], RankOfObject( a ), RankOfObject( b ), Q ), b );;
gap> beta := CategoryOfRowsMorphism( c, HomalgMatrix( [ 8 .. 31 ], RankOfObject( c ), RankOfObject( d ), Q ), d );;
gap> gamma := CategoryOfRowsMorphism( a_tensor_b, HomalgMatrix( [ 2 .. 7 ], RankOfObject( a_tensor_b ), RankOfObject( u ), Q ), u );;
gap> delta := CategoryOfRowsMorphism( c_tensor_d, HomalgMatrix( [ 8 .. 31 ], RankOfObject( c_tensor_d ), RankOfObject( u ), Q ), u );;
gap> epsilon := CategoryOfRowsMorphism( u, HomalgMatrix( [ 2 .. 7], RankOfObject( u ), RankOfObject( hom_ab ), Q ), hom_ab );;
gap> zeta := CategoryOfRowsMorphism( u, HomalgMatrix( [ 8 .. 31 ], RankOfObject( u ), RankOfObject( hom_cd ), Q ), hom_cd );;
gap> 
gap> 
gap> StartTimer( "without" );;
gap> ClosedMonoidalCategoriesTest( rows, a, b, c, d, alpha, beta, gamma, delta, epsilon, zeta );;
gap> StopTimer( "without" );;
gap> DisplayTimer( "without" );;
Timer without (stopped): started 1 times with a total runtime of 1533 ms ( = 1533000 μs per run)
gap> 
gap> 
gap> StartTimer( "primitive" );;
gap> ClosedMonoidalCategoriesTest( rows, a, b, c, d, alpha, beta, gamma, delta, epsilon, zeta : only_primitive_operations := true );;
gap> StopTimer( "primitive" );;
gap> DisplayTimer( "primitive" );;
Timer primitive (stopped): started 1 times with a total runtime of 3 ms ( = 3000 μs per run)
```
Now, running the test functions in reversed order in a new GAP session:

```gap
gap> LoadPackage("freyd");
gap> Q := HomalgFieldOfRationals();;
gap> rows := CategoryOfRows( Q );;
gap> u := TensorUnit( rows );;
gap> 
gap> a := 2 / rows;;
gap> b := 3 / rows;;
gap> c := 6 / rows;;
gap> d := 4 / rows;;
gap> 
gap> a_tensor_b := TensorProduct( a, b );;
gap> c_tensor_d := TensorProduct( c, d );;
gap> 
gap> hom_ab := InternalHom( a, b );;
gap> hom_cd := InternalHom( c, d );;
gap> 
gap> alpha := CategoryOfRowsMorphism( a, HomalgMatrix( [ 2 .. 7 ], RankOfObject( a ), RankOfObject( b ), Q ), b );;
gap> beta := CategoryOfRowsMorphism( c, HomalgMatrix( [ 8 .. 31 ], RankOfObject( c ), RankOfObject( d ), Q ), d );;
gap> gamma := CategoryOfRowsMorphism( a_tensor_b, HomalgMatrix( [ 2 .. 7 ], RankOfObject( a_tensor_b ), RankOfObject( u ), Q ), u );;
gap> delta := CategoryOfRowsMorphism( c_tensor_d, HomalgMatrix( [ 8 .. 31 ], RankOfObject( c_tensor_d ), RankOfObject( u ), Q ), u );;
gap> epsilon := CategoryOfRowsMorphism( u, HomalgMatrix( [ 2 .. 7], RankOfObject( u ), RankOfObject( hom_ab ), Q ), hom_ab );;
gap> zeta := CategoryOfRowsMorphism( u, HomalgMatrix( [ 8 .. 31 ], RankOfObject( u ), RankOfObject( hom_cd ), Q ), hom_cd );;
gap> 
gap> 
gap> StartTimer( "primitive" );;
gap> ClosedMonoidalCategoriesTest( rows, a, b, c, d, alpha, beta, gamma, delta, epsilon, zeta : only_primitive_operations := true );;
gap> StopTimer( "primitive" );;
gap> DisplayTimer( "primitive" );;
Timer primitive (stopped): started 1 times with a total runtime of 121192 ms ( = 121192000 μs per run)
gap> 
gap> 
gap> StartTimer( "without" );;
gap> ClosedMonoidalCategoriesTest( rows, a, b, c, d, alpha, beta, gamma, delta, epsilon, zeta );;
gap> StopTimer( "without" );;
gap> DisplayTimer( "without" );;
Timer without (stopped): started 1 times with a total runtime of 9 ms ( = 9000 μs per run)
```
So the order of the calls to the test functions makes a huge difference. :/

**Comment by zickgraf** [comment]
> So the order of the calls to the test functions makes a huge difference. :/

Ah, that's expected: `Opposite` is an attribute, so `only_primitive_operations` will not be respected if you have already called the function without it, and the tests will just return the already cached results. Good thing that we look at this more closely!

**Comment by zickgraf** [comment]
I have found the culprit: The associativity of `PreCompose`. In the example, matrices of the following sizes are multiplied in `CoevaluationMorphismWithGivenRange/CoclosedCoevaluationMorphismWithGivenSource` (due to op one has to think transposed):
```
gap> UnderlyingMatrix(Opposite(mor1));
<An unevaluated 82944 x 82944 identity matrix over an internal ring>
gap> UnderlyingMatrix(Opposite(mor2));
<A 82944 x 82944 matrix over an internal ring>
gap> UnderlyingMatrix(Opposite(mor3));
<A 82944 x 82944 matrix over an internal ring>
gap> UnderlyingMatrix(Opposite(mor4));
<A 144 x 82944 matrix over an internal ring>
```

If we multiply these matrices starting with `mor3` and `mor4`, then we get "small" 144 x 82944 matrices in every step. If we start with `mor1` and `mor2`, we get 82944 x 82944 matrices as intermediate results.

So if one replaces `PreComposeList` by `PostComposeList` (and swaps the arguments) in `CoclosedCoevaluationMorphismWithGivenSource`, then the code will be fast.

Two thoughts:
1. We could discuss if `mor4` is always small, so it always makes sense to start with `mor4`. If not, we cannot really optimize in general.
2. Maybe the dual versions of the derivations should always use `PostComposeList` instead of swapping the arguments. This would have the additional benefit that the two versions are easier to compare, because swapping the arguments creates a large diff.

**Comment by zickgraf** [comment]
> Did you use the compatiblity morphisms again or could you give me more details on what you executed with which parameters ? Which derivations of CoevaluationMorphismWithGivenRange/CoclosedCoevaluationMorphismWithGivenSource
was finally triggered?

Sorry, those remarks were meant for further verbal discussion, not for the full picture :D Yes, I started with the same arguments and `InternalCoHomTensorProductCompatibilityMorphism` as before, which runs into the derivations you mention. With `mor1`, ..., `mor4` I mean the four morphisms occurring in the `PreComposeList` in  `CoclosedCoevaluationMorphismWithGivenSource`. So the question is: Is
```
TensorProductOnMorphisms( cat, 
    CoclosedCoevaluationForCoDual( cat, b ), 
    id_a )
```
always "small" compared to the other morphisms?

> On a completely different note: there are algorithms to be smarter about multiplying a list of matrices (e.g. https://www.eti.uni-siegen.de/ti/lehre/ws2021/algo1/alg1.pdf page 160-165)

Yes, but of course this only works heuristically (e.g. concretely given sparse matrices might behave completely different).

> but I don't understand the transition from CAP to homalg good enough, to see if this could help.

It could: `PreComposeList` would have to be an official CAP operation, and we would need a `Product` function for matrices in MatricesForHomalg, which `PreComposeList` in MatrixCategory could call. Then homalg would get a list of matrices and could try to do optimizations on it.

> Many of the derivations are also written upside-down from their dual counterparts, so this might make sense. However, I think I tried that originally and got confused because PostCompose is harder to read for a large list.
If that turns out to be faster/better and I have to touch a large amount of derivations again, I'll be seriously annoyed. :D

We can discuss this on Monday, but I think I'm in favor of `PostComposeList`, sorry :D

**Comment by TKuh** [comment]
I used the following code to play around a bit:

```gap
test := function( cat, a, b ) 
    local morphism, codual_b, id_a, mor1, mor2, mor3, mor4;
     
    codual_b := CoDualOnObjects( cat, b ); 
          
    id_a := IdentityMorphism( cat, a ); 

    mor1 := IsomorphismFromInternalCoHomToTensorProduct( cat, 
                        TensorProductOnObjects( cat, a, b ), 
                        b);
                        
    mor2 := TensorProductOnMorphisms( cat, 
                        IdentityMorphism( cat, codual_b ), 
                        Braiding( cat, a, b ) );
                        
    mor3 := TensorProductOnMorphisms( cat, 
                        Braiding( cat, codual_b, b ), 
                        id_a );
                        
    mor4 := TensorProductOnMorphisms( cat, 
                        CoclosedCoevaluationForCoDual( cat, b ), 
                        id_a );
                  
    Print( "mor1: ", NrRows( UnderlyingMatrix( mor1 ) ) , "x", NrColumns( UnderlyingMatrix( mor1 ) ), "\n" );
    
    Print( "mor2: ", NrRows( UnderlyingMatrix( mor2 ) ) , "x", NrColumns( UnderlyingMatrix( mor2 ) ), "\n" );
    
    Print( "mor3: ", NrRows( UnderlyingMatrix( mor3 ) ) , "x", NrColumns( UnderlyingMatrix( mor3 ) ), "\n" );
    
    Print( "mor4: ", NrRows( UnderlyingMatrix( mor4 ) ) , "x", NrColumns( UnderlyingMatrix( mor4 ) ), "\n" );

end;
```

And it seems you are right, `mor4` is always the smallest one:

```gap
gap> test(rows, 2/rows, 4/rows);
mor1: 32x32
mor2: 32x32
mor3: 32x32
mor4: 32x2
gap> test(rows, 2/rows, 5/rows);
mor1: 50x50
mor2: 50x50
mor3: 50x50
mor4: 50x2
gap> test(rows, 2/rows, 6/rows);
mor1: 72x72
mor2: 72x72
mor3: 72x72
mor4: 72x2
gap> test(rows, 3/rows, 4/rows);
mor1: 48x48
mor2: 48x48
mor3: 48x48
mor4: 48x3
gap> test(rows, 3/rows, 5/rows);
mor1: 75x75
mor2: 75x75
mor3: 75x75
mor4: 75x3
gap> test(rows, 3/rows, 6/rows);
mor1: 108x108
mor2: 108x108
mor3: 108x108
mor4: 108x3
gap> test(rows, 4/rows, 5/rows);
mor1: 100x100
mor2: 100x100
mor3: 100x100
mor4: 100x4
gap> test(rows, 4/rows, 6/rows);
mor1: 144x144
mor2: 144x144
mor3: 144x144
mor4: 144x4
gap> test(rows, 5/rows, 6/rows);
mor1: 180x180
mor2: 180x180
mor3: 180x180
mor4: 180x5
gap> test(rows, 6/rows, 6/rows);
mor1: 216x216
mor2: 216x216
mor3: 216x216
mor4: 216x6
gap> test(rows, 4/rows, 2/rows);
mor1: 16x16
mor2: 16x16
mor3: 16x16
mor4: 16x4
gap> test(rows, 4/rows, 3/rows);
mor1: 36x36
mor2: 36x36
mor3: 36x36
mor4: 36x4
gap> test(rows, 5/rows, 3/rows);
mor1: 45x45
mor2: 45x45
mor3: 45x45
mor4: 45x5
gap> test(rows, 5/rows, 4/rows);
mor1: 80x80
mor2: 80x80
mor3: 80x80
mor4: 80x5
gap> test(rows, 6/rows, 3/rows);
mor1: 54x54
mor2: 54x54
mor3: 54x54
mor4: 54x6
gap> test(rows, 6/rows, 4/rows);
mor1: 96x96
mor2: 96x96
mor3: 96x96
mor4: 96x6
gap> test(rows, 6/rows, 5/rows);
mor1: 150x150
mor2: 150x150
mor3: 150x150
mor4: 150x6
gap> test(rows, 6/rows, 6/rows);
mor1: 216x216
mor2: 216x216
mor3: 216x216
mor4: 216x6

```

And in fact one will always have `dim Range(mor4) = dim a` since the range of `CoclosedCoevaluationForCoDual` is the tensor unit. And any other morphism has some sort of tensor product of `a` and `b` in the source and range.
